src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,161 @@
+#include "reactor-uc/platform/riot/uart_channel.h"
+#include "reactor-uc/logging.h"
+#include "reactor-uc/environment.h"
+#include "reactor-uc/serialization.h"
+#include "led.h"
+
+#define UART_CHANNEL_ERR(fmt, ...) LF_ERR(NET, "UARTSyncChannel: " fmt, ##__VA_ARGS__)
+#define UART_CHANNEL_WARN(fmt, ...) LF_WARN(NET, "UARTSyncChannel: " fmt, ##__VA_ARGS__)
+#define UART_CHANNEL_INFO(fmt, ...) LF_INFO(NET, "UARTSyncChannel: " fmt, ##__VA_ARGS__)
+#define UART_CHANNEL_DEBUG(fmt, ...) LF_DEBUG(NET, "UARTSyncChannel: " fmt, ##__VA_ARGS__)
+
+static lf_ret_t UARTSyncChannel_open_connection(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Open connection");
+  (void)untyped_self;
+  return LF_OK;
+}
+
+static void UARTSyncChannel_close_connection(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Close connection");
+  (void)untyped_self;
+}
+
+static void UARTSyncChannel_free(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Free");
+  (void)untyped_self;
+}
+
+static bool UARTSyncChannel_is_connected(NetworkChannel *untyped_self) {
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+  return self->state == NETWORK_CHANNEL_STATE_CONNECTED;
+}
+
+static lf_ret_t UARTSyncChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+
+  if (self->state == NETWORK_CHANNEL_STATE_CONNECTED) {
+    int message_size = serialize_to_protobuf(message, self->write_buffer, UART_CHANNEL_BUFFERSIZE);
+
+    UART_CHANNEL_DEBUG("Sending Message of Size: %i", message_size);
+
+    uart_write(self->uart_dev, self->write_buffer, message_size);
+    return LF_OK;
+  } else {
+    return LF_ERR;
+  }
+}
+
+static void UARTSyncChannel_register_receive_callback(NetworkChannel *untyped_self,
+                                                      void (*receive_callback)(FederatedConnectionBundle *conn,
+                                                                               const FederateMessage *msg),
+                                                      FederatedConnectionBundle *conn) {
+  UART_CHANNEL_INFO("Register receive callback");
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+void _UARTSyncChannel_receive_callback(void *arg, uint8_t received_byte) {
+  UARTSyncChannel *self = (UARTSyncChannel *)arg;
+  const uint32_t minimum_message_size = 12;
+
+  self->read_buffer[self->read_index] = received_byte;
+  self->read_index++;
+
+  if (self->read_index >= minimum_message_size) {
```

Comment:
It is a little unfortunate that we will try decoding for each byte we receive. It suggests that we may want some framing around the message, i.e. a header with a length field. But that will be future work

Reply:
Or just attempt decoding every X received bytes.

Reply:
Yes, but that could lead to unexpected results if you receive messages very rarely. Imagine you have received the entire message, but you have received X-1 bytes since the last decode attempt. So we would also need to take time into account.

Reply:
also true

Reply:
lets change this when we converged in the discussion #199